### PR TITLE
Expressions with null

### PR DESCRIPTION
**Pull request by kkondaka**
### Description
Added support for checking for "null" in the expressions.
With this change expressions like `/response == null` or `/response != null` can be specified as shown below

```
 processor:
     - grok:
        match:
          message: ['%{IPORHOST:clientip} (?:%{WORD:ident}|-) (%{USER:auth}|-) \[%{HTTPDATE:timestamp}\] "(?:%{WORD:verb} %{NOTSPACE:request}(?: HTTP/%{NUMBER:httpversion})?|%{DATA:rawrequest})" %{NUMBER:response} (?:%{NUMBER:bytes}|-) %{BASE10NUM:timetaken} "(?:%{WORD:ref}|-)" %{GREEDYDATA:useragent}']
    - drop_events:
         drop_when '/request == null'
 ```
### Issues Resolved
[List any issues this PR will resolve]
#1136 
 
### Check List
- [X ] New functionality includes testing.
- [X ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java [hunk 1]
```diff
@@ -42,6 +42,8 @@ public Object coercePrimaryTerminalNode(final TerminalNode node, final Event eve
                 return Float.valueOf(nodeStringValue);
             case DataPrepperExpressionParser.Boolean:
                 return Boolean.valueOf(nodeStringValue);
+            case DataPrepperExpressionParser.Null:
+                return null;
```
Comment (dlvenable) [inline data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java:46]: I like that this works out of box with the operators. But, should this be some sort of special class - like `NullValue`? I suppose this could come later if that provides specific value.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ConditionalExpressionEvaluatorIT.java [hunk 2]
```diff
@@ -129,6 +129,9 @@ private static Stream<Arguments> validExpressionArguments() {
                 Arguments.of("true == (/is_cool == true)", event("{\"is_cool\": true}"), true),
                 Arguments.of("not /is_cool", event("{\"is_cool\": true}"), false),
                 Arguments.of("/status_code < 300", event("{\"status_code\": 200}"), true),
+                Arguments.of("/status_code != null", event("{\"status_code\": 200}"), true),
+                Arguments.of("/response == null", event("{\"status_code\": 200}"), true),
+                Arguments.of("/response != null", event("{\"status_code\": 200}"), false),
```
[the remark this answers is not in the document]
Comment (sbayer55) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ConditionalExpressionEvaluatorIT.java:134]: I recommend adding a few test cases with null on the left had side to test operators do not throw unexpected exceptions.

An example where a bug could occur:
Given a hypothetical operator `foo` that checks for class equality implemented as `(lhs, rhs) -> lhs.getClass().equals(rhs.getClass())`

When invoked `null foo null`

Then a `NullPointerException` will be thrown instead of `true`

**Review by sbayer55 — CHANGES_REQUESTED** [review]
Thanks for adding `null` value support. It's cool to see Data Prepper Expressions syntax grow.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java [hunk 3]
```diff
@@ -42,6 +42,8 @@ public Object coercePrimaryTerminalNode(final TerminalNode node, final Event eve
                 return Float.valueOf(nodeStringValue);
             case DataPrepperExpressionParser.Boolean:
                 return Boolean.valueOf(nodeStringValue);
+            case DataPrepperExpressionParser.Null:
```
[the remark this answers is not in the document]
Comment (kkondaka) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java:45]: Yes, I thought that's what the change in this file - data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ParseTreeEvaluatorListenerTest.java? Is there another unit test file?

File: docs/expression_syntax.md [hunk 4]
```diff
@@ -83,6 +83,8 @@ Examples
 /status_code == 200 and /message == "Hello world"
 /status_code == 200 or /status_code == 202
 not /status_code in {200, 202}
+/response == null
+/response != null
```
[the remark this answers is not in the document]
Comment (kkondaka) [reply docs/expression_syntax.md:87]: I have added a statement in the "literals" section. `- Null _(Supports null check to see if a Json Pointer is present or not)_`. Is that not clear enough?

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ConditionalExpressionEvaluatorIT.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kkondaka) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ConditionalExpressionEvaluatorIT.java:134]: I can add that. I see no harm in adding more tests.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java (hunk 1 above)
Comment (kkondaka) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java:46]: Frankly, I have no idea. I will follow what you suggest here.

File: docs/expression_syntax.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkondaka) [reply docs/expression_syntax.md:102]: Added more documentation with examples.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkondaka) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java:45]: Added a unit test in data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ParseTreeCoercionServiceTest.java

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ConditionalExpressionEvaluatorIT.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kkondaka) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ConditionalExpressionEvaluatorIT.java:137]: Added few test cases with null on the left side.

File: data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeCoercionService.java:46]: It might not matter. And we can change it later since these are internal classes.

**Review by dlvenable — APPROVED** [review]
Thanks for expanding the conditional support with this! And thanks for making the requested test changes to help ensure quality!

File: data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ConditionalExpressionEvaluatorIT.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbayer55) [reply data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/ConditionalExpressionEvaluatorIT.java:137]: Thanks!

**Review by sbayer55 — APPROVED** [review]
